src/utils/cloudflare/api.js

```diff
@@ -0,0 +1,114 @@
+const assert = require('assert');
+const retry = require('promise-retry');
+const { helpers: { generateClass } } = require('common-errors');
+
+const { CloudflareClient } = require('./client');
+
+const RULE_LIST = 'accounts/:accountId/rules/lists';
+const RULE_ITEM_LIST = 'accounts/:accountId/rules/lists/:listId/items';
+const RULE_BATCH_PROGESS = 'accounts/:accountId/rules/lists/bulk_operations/:operationId';
+
+const BulkOperationError = generateClass('BulkOperationError', {
+  args: ['message', 'response'],
+});
+
+const defaultRetryConfig = {
+  retries: 10,
+  factor: 0.5,
+};
+
+class CloudflareAPI {
+  constructor(cfClient, config = {}, withProps = {}) {
+    assert(cfClient instanceof CloudflareClient, 'invalid client');
+    this.config = config;
+    this.cfClient = cfClient;
+    this.http = cfClient.getHttpClient();
+    this.props = withProps;
+  }
+
+  withAccountId(accountId) {
+    return new CloudflareAPI(this.cfClient, this.config, { accountId });
+  }
+
+  apiUrl(t, values) {
+    const { accountId } = this.props;
+    assert(accountId, 'accountId invalid');
+    return Object.entries({ accountId, ...values })
+      .reduce((prev, [prop, val]) => prev.replace(`:${prop}`, val), t);
+  }
```

Comment:
this is an expensive operation (iterating, searching, replacing) - consider using template literals (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)

Reply:
Changed to template literals + extra Fns.